### ISSUE TITLE
AKS support VNet IPv6 addresses using CNI

### ISSUE THREAD

**Issue by CpuID** [Azure/AKS#2732]
As a follow-up to https://github.com/Azure/AKS/issues/460 (since comments were locked) - I see that kubenet support has gone into preview recently.

Are there plans to support CNI IPv6 addresses also with AKS? eg. pods get their own unique VNet IPv6 address, the same as they do with IPv4 when using CNI today?

**Comment by CpuID** [comment]
@danimal521 @phealy @palma21 @tyler-lloyd ^^

**Comment by phealy** [comment]
Yes, it's planned. Currently [you cannot add more than one private IPv6 address to a NIC in Azure](https://docs.microsoft.com/en-us/azure/virtual-network/ip-services/virtual-network-network-interface-addresses#:~:text=existing%20network%20interface.-,Each%20network%20interface%20may%20have%20at%20most%20one%20IPv6%20private%20address.,-You%20can%20optionally), a limitation at the networking layer. As Azure CNI requires adding secondary IP addresses to the NIC for pods, we cannot bring IPv6 to Azure CNI until this limitation is resolved by the networking team. We have requested this but I do not have a date I can share at this time.

**Comment by CpuID** [comment]
ok thanks @phealy 👍 mind keeping this open for updates as they become available in future?

**Comment by phealy** [comment]
No, I'll convert it to the roadmap item tomorrow. Just didn't have time today.

**Comment by CpuID** [comment]
no probs thx

**Comment by CpuID** [comment]
Hey @phealy just wanted to check did this make it onto the roadmap? I know you're still waiting on the networking team to get prefix/multi IP v6 support on the NIC's first.

Checked in this repo under Projects but couldn't see any specific reference to it (other than the kubenet IPv6 item)?

**Comment by tkuntal** [comment]
Hello @phealy we are currently using Azure CNI networking. We are onboarding to Private Endpoint as part of our public preview requirement. For this we need dual stack enabled deployment and moving to dual stack would imply having to use kubenet networking which is like a step backward for us. So just wanted to check where this feature is in your pipeline?

**Comment by CpuID** [comment]
hey @phealy can you get this on the roadmap? :) can't see it there yet. much appreciated

**Comment by agronholm** [comment]
> Yes, it's planned. Currently [you cannot add more than one private IPv6 address to a NIC in Azure](https://docs.microsoft.com/en-us/azure/virtual-network/ip-services/virtual-network-network-interface-addresses#:~:text=existing%20network%20interface.-,Each%20network%20interface%20may%20have%20at%20most%20one%20IPv6%20private%20address.,-You%20can%20optionally), a limitation at the networking layer. As Azure CNI requires adding secondary IP addresses to the NIC for pods, we cannot bring IPv6 to Azure CNI until this limitation is resolved by the networking team. We have requested this but I do not have a date I can share at this time.

Is there an open issue we could follow?

**Comment by sparkdevo** [comment]
This feature is so useful that our migration plan is blocked because of it. @phealy

**Comment by ewfian** [comment]
same issue here

**Comment by paulgmiller** [comment]
Geting vnet ips that are ipv6 is still a ways out.  And I am not sure it will solve ip exhaustion because we only have dualstack not single ipv6 stack. 
We are getting closer to ipv6 in [azure overlay ](https://learn.microsoft.com/en-us/azure/aks/azure-cni-overlay) which helps with ip exhaustion and scales very well but won't help if your pods need to be directly addressable from outside the vnet.

Very curious how many of the people on this thread need directly addressable ipv6 pods.

**Comment by agronholm** [comment]
The ideal for me would be IPv6-only internal network, with external IPv4/IPv6 connectivity via load balancer. I take it this is not yet possible?

**Comment by etaham** [comment]
Also curious if this will raise the 250 pod limit per host? That is limiting our density/efficiency.

> Geting vnet ips that are ipv6 is still a ways out. And I am not sure it will solve ip exhaustion because we only have dualstack not single ipv6 stack. We are getting closer to ipv6 in [azure overlay ](https://learn.microsoft.com/en-us/azure/aks/azure-cni-overlay) which helps with ip exhaustion and scales very well but won't help if your pods need to be directly addressable from outside the vnet.
> 
> Very curious how many of the people on this thread need directly addressable ipv6 pods.

We don't need directly addressable POD - exposing as services would work.

**Comment by jkroepke** [comment]
> Also curious if this will raise the 250 pod limit per host? That is limiting our density/efficiency.

Kubernetes is designed for 110 pod/node. https://kubernetes.io/docs/setup/best-practices/cluster-large/. 250 is already far behind the design spec.

The 250 pod limit depends on the node network size, which is /24 by default and not configurable yet.

**Comment by etaham** [comment]
If I'm reading this correctly, it looks like EKS supports a lot more: https://github.com/awslabs/amazon-eks-ami/blob/master/files/eni-max-pods.txt. Using IPV6 for the POD network would likely allow for a much larger IP block allocation per host without concern of exhaustion.

High node density enables higher scale services deployments. The kubelet's capacity for amount of PODs per node will be a function of what the POD lifecycle, unique to each use case. There is little work to do for pods that live for many hours/days. If you have PODs that live for only a few seconds, the kubelet will under much more load and stress when trying to match the same number of PODs.

**Comment by jkroepke** [comment]
> it looks like EKS supports a lot more:

110 max as scaling spec defines it and 250 max if you node has more than 30 vCPUs. Same maximum as Azure.

https://github.com/awslabs/amazon-eks-ami/blob/487830668b1f556e530fd1d6310763b337cfb74e/files/max-pods-calculator.sh#L145-L147


> High node density enables higher scale services deployments. 

I understand you use case, however Kubernetes is not designed for that.

**Comment by aL3891** [comment]
Any news on this? it would make things alot easier for us (or allowing windows nodes on kubenet clusters)

**Comment by paulgmiller** [comment]
Likely going to close this open more targeted items around ipv6
1) Azure CNI overlay does support dual stack.  This unblocks windows dualstack and a few other things
https://learn.microsoft.com/en-us/azure/aks/azure-cni-overlay?tabs=kubectl#deploy-a-dual-stack-aks-cluster

2) Azure CNI nodesubnet and podsubnet (with in vnet ips) does not support dual stack yet and is on hold. 

3) AKS does not have an ipv6 single stack solution at all yet. Really curious how much demand there is for that so opening an separate item.

**Comment by paulgmiller** [comment]
podsubnet dual stack: https://github.com/Azure/AKS/issues/4389

ipv6 single stack:  https://github.com/Azure/AKS/issues/4388
